src/frontend/contexts/LocalDiscoveryContext.tsx

```diff
@@ -319,6 +375,35 @@ function stopZeroconf(zeroconf: Zeroconf): Promise<void> {
   });
 }
 
+function unpublishZeroconf(
+  zeroconf: Zeroconf,
+  publishedNames: Set<string>,
+): Promise<void> {
+  if (publishedNames.size === 0) return Promise.resolve();
+  return new Promise((resolve, reject) => {
+    const timeoutId = setTimeout(() => {
+      cleanup();
+      reject(new Error('Timed out unpublishing zeroconf service'));
+    }, ZEROCONF_UNPUBLISH_TIMEOUT_MS);
+
+    const cleanup = () => {
+      clearTimeout(timeoutId);
+      zeroconf.off('remove', onRemove);
+    };
+    const onRemove = (name: string) => {
+      publishedNames.delete(name);
```

Comment:
I was pretty surprised to see this mutate `publishedNames`. We can avoid that by returning a result type:

```javascript
function unpublishZeroconf(
  zeroconf: Zeroconf,
  publishedNames: Readonly<Set<string>>,
): Promise<{
  publishedNames: Set<string>;
  err?: Error;
}> {
  if (publishedNames.size === 0) return Promise.resolve({publishedNames});
  publishedNames = new Set(publishedNames);
  return new Promise(resolve => {
    const timeoutId = setTimeout(() => {
      cleanup();
      resolve({
        publishedNames,
        err: new Error('Timed out unpublishing zeroconf service'),
      });
    }, ZEROCONF_UNPUBLISH_TIMEOUT_MS);

    const cleanup = () => {
      clearTimeout(timeoutId);
      zeroconf.off('remove', onRemove);
    };
    const onRemove = (name: string) => {
      publishedNames.delete(name);
      if (publishedNames.size === 0) {
        cleanup();
        resolve({publishedNames});
      }
    };
    zeroconf.on('remove', onRemove);
    for (const name of publishedNames) {
      zeroconf.unpublishService(name);
    }
  });
}
```

We can then use it above like this:

```javascript
// ...
(async () => {
  const unpublishResult = await unpublishZeroconf(
    zeroconf,
    publishedNames,
  );
  ({publishedNames} = unpublishResult);
  if (unpublishResult.err) {
    // See above for why we silently fail here
    Sentry.captureException(unpublishResult.err);
  }
})(),
// ...
```

More verbose, but I think it avoids a pretty unusual mutation of the argument.

(You could argue that we're mutating `zeroconf` too, but I feel like that's a bit different; that's passing a handle to some external interface, which the set is not.)

Reply:
Not such a fan of this suggestion. `publishedNames` is really tracking state that should be internal to zeroconf. It feels weird that it gets "out of sync" if it's not mutated - unpublish is async, so changes to this code to read `publishedNames` before unpublish resolves would not get the correct answer. We don't currently have code that would be affected by this, and we might not, but it seems like we're opening ourselves to a race condition. I think because `publishedNames` is the state of "which names are published", then it should be mutated as names are published / unpublished. If there is an error during the unpublishing of one of the names, we don't want the unpublishign of the other names to be ignored / missing from state.
TL;DR I think this change will introduce bugs, and IIFEs can also be hard to read.

Reply:
Renamed the arg in 5cb2e2fc808da2d28e4258967fd1d583f8f38f3a.